src/workerd/io/actor-sqlite.c++

```diff
@@ -169,11 +200,67 @@ void ActorSqlite::onWrite() {
       // rather than after the callback.
       { auto drop = kj::mv(txn); }
 
-      return commitCallback();
+      return commitImpl(kj::mv(precommitAlarmPromise));
     })));
   }
 }
 
+kj::Promise<void> ActorSqlite::requestScheduledAlarm(kj::Maybe<kj::Date> requestedTime) {
+  co_await hooks.scheduleRun(requestedTime);
+  lastConfirmedScheduledAlarm = requestedTime;
+}
+
+ActorSqlite::PrecommitAlarmState ActorSqlite::startPrecommitAlarmScheduling() {
+  PrecommitAlarmState state;
+  state.localAlarmState = metadata.getAlarm();
+  if (pendingCommit == kj::none &&
+      willFireEarlier(state.localAlarmState, lastConfirmedScheduledAlarm)) {
+    // Basically, this is the first scheduling request that commitImpl() would make prior to
+    // commitCallback().  We start the request separately, ahead of calling sqlite functions that
+    // commit to local disk, for correctness in workerd, where alarm scheduling and db commits are
+    // both synchronous.
+    state.schedulingPromise = requestScheduledAlarm(state.localAlarmState);
+  }
+  return kj::mv(state);
+}
+
+kj::Promise<void> ActorSqlite::commitImpl(ActorSqlite::PrecommitAlarmState precommitAlarmState) {
+  // We assume that exceptions thrown during commit will propagate to the caller, such that they
+  // will ensure cancelDeferredAlarmDeletion() is called, if necessary.
+
+  KJ_IF_SOME(pending, pendingCommit) {
+    co_await pending.addBranch();
+    co_return;
```

Comment:
I guess we're potentially canceling `precommitAlarmState.schedulingPromise` here -- and worse, it may or may not have actually completed.

But I think this is OK, because the commit we're merging with will have executed the `while` loop below and accounted for any alarm change required by our new commit?

Probably worth commenting to point this out.

Reply:
I think it shouldn't actually cancel schedulingPromise, since `startPrecommitAlarmScheduling()` currently does not issue a scheduling request and populate the promise, if it sees that pendingCommit is already populated -- I assume that's OK?  But I can add more comments to clarify the reasoning.

Reply:
Added more documentation to commitImpl() in 784a55365 and d1983124b